### PR TITLE
[cmake] linux: add option to enable link time optimization

### PR DESCRIPTION
**Pull request by wsnipex**
Allow building with LTO: https://gcc.gnu.org/onlinedocs/gcc-7.2.0/gccint/LTO-Overview.html#LTO-Overview

## Description
add new USE_LTO option
optionally specify USE_LTO=[int] for parallel execution of linker jobs (just like make -j[int])

## Motivation and Context

better compiler optimizations.
Did not yet really test performance differences, but resulting kodi binary is ~15% smaller

## How Has This Been Tested?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: cmake/scripts/linux/ArchSetup.cmake [hunk 1]
```diff
@@ -37,9 +37,31 @@ else()
   endif()
 endif()
 
-# Make sure we strip binaries in Release build
 if(CMAKE_BUILD_TYPE STREQUAL Release AND CMAKE_COMPILER_IS_GNUCXX)
```
Comment (MilhouseVH) [inline cmake/scripts/linux/ArchSetup.cmake]: Would it be possible to support `CMAKE_BUILD_TYPE STREQUAL MinSizeRel` as well?

Comment (wsnipex) [reply cmake/scripts/linux/ArchSetup.cmake]: added

**Comment by MilhouseVH** [comment]
What follows is results of some analysis I did yesterday when building LibreELEC for Generic with this PR. Currently we strip the LTO flags from the CFLAGS using a function called [`strip_lto()`](https://github.com/LibreELEC/LibreELEC.tv/blob/efdc8f64b735ebb73521ca1b3f59978324f0c9e2/config/functions#L440-L452) and we call this from the kodi build [package](https://github.com/LibreELEC/LibreELEC.tv/blob/efdc8f64b735ebb73521ca1b3f59978324f0c9e2/packages/mediacenter/kodi/package.mk#L245-L246).

tl;dr: This PR works well, and saves about 2MB (23MB to 21MB) on an x86_64 build, and about 1MB (19MB to 18MB) on an arm (RPi2) build. My test builds (starting with #1210) are now based on this PR (and the branch listed at the foot of this post).

---

USE_LTO results for Generic (build host: FX-8350 with 8 cores):

```
Test# | USE_LTO=8 | strip_lto() |  Size Delta | Avg Time
------+-----------+-------------+-------------+----------
  1   |   NO      |    YES      |           0 | 249 secs        
  2   |   YES     |    NO       |  -2,121,112 | 278 secs
  3   |   YES     |    YES      |  -2,121,112 | 264 secs
  4   |   NO      |    NO       |  -2,121,112 | 530 secs
```

Test #1 is the baseline spec for current master, ie. no LTO at all.
Test #2 is: `USE_LTO` enabled, remove `strip_lto()` from package.mk
Test #3 is: `USE_LTO` enabled, but continue using `strip_lto()` (removing LTO from CFLAGS)
Test #4 is: `USE_LTO` disabled, relying only on CFLAGS for single-threaded LTO support
`Size Delta`: kodi.bin size relative to that from Test #1
`Avg Time`: time in seconds over 4 runs building `kodi` and creating a tarball (ie. includes mksquash overhead)

1 is fastest (as LTO is completely disabled) but results in a larger file size
2, 3 & 4 all use some form of LTO, resulting in a similar size reduction
2 & 3 are almost as fast as 1 due to parallel linking. Average time will vary depending on number of cores
4 is by far the slowest, due to single core LTO (via CFLAGS). Presumably this is why LTO has always been disabled?

Observations:

1. Stripping or not stripping LTO from CFLAGS still has an impact even when using cmake
2. Non-parallel LTO (test 4) increases link time significantly
3. Benefits of `USE_LTO` will most likely vary with number of cores/CPUs available
4. `USE_LTO` will only work for `Release` and `MinSizeRel` builds - we still need `strip_lto()` for DEBUG builds
5. Similar results on RPi2 (file size reduced from 19,925,536 to 18,975,352)

Recommendation (IMHO):

1. Use "Test #3" configuration
2. Enable `-DUSE_LTO=$CONCURRENCY_MAKE_LEVEL` (optional, if `$LTO_SUPPORT` is enabled)
3. Continue using `strip_lto()` in `kodi/package.mk`

Branch: https://github.com/LibreELEC/LibreELEC.tv/compare/master...MilhouseVH:le90_use_lto

**Comment by wsnipex** [comment]
thanks for the detailed test

**Comment by notspiff** [comment]
i experimented with LTO support back when i did the cmake system and you could not successfully link kodi with lto in the gcc 4(.6/8) days. that's why i never added it at least.

**Comment by candrews** [comment]
Why not always do:
`set(CMAKE_INTERPROCEDURAL_OPTIMIZATION TRUE)` ?

I don't think there's any downside.
